security/src/main/java/com/fincity/security/service/UserService.java

```diff
@@ -652,54 +653,62 @@ public Mono<Boolean> updateNewPassword(String urlAppCode, String urlClientCode,
-		}
+				                passwordEqual -> passwordEqual.booleanValue() ? Mono.just(passwordEqual)
+				                        : securityMessageResourceService.throwMessage(
 
-		return flatMapMonoWithNull(
+				                                msg -> new GenericException(HttpStatus.FORBIDDEN, msg),
+				                                SecurityMessageResourceService.OLD_NEW_PASSWORD_MATCH))
+				                .contextWrite(Context.of(LogUtil.METHOD_NAME, "UserService.checkHierarchy"))
+				                .log();
 
-		        SecurityContextUtil::getUsersContextAuthentication,
+			        }
 
-		        ca -> ca.isSystemClient() ? Mono.justOrEmpty(Optional.empty())
-		                : this.dao.readById(reqUserId)
-		                        .map(User::getClientId),
+			        return flatMapMono(
 
-		        (ca, rClientId) -> Mono.just(ca.isSystemClient() || rClientId.toBigInteger()
-		                .equals(ca.getUser()
-		                        .getClientId())),
+			                () -> Mono.just(ca.isSystemClient() || user.getClientId()
+			                        .equals(loggedInUserClientId)),
 
-		        (ca, rClientId, isSameClient) -> isSameClient.booleanValue() ? Mono.just(true)
-		                : this.clientService.isBeingManagedBy(ULong.valueOf(ca.getUser()
-		                        .getClientId()), rClientId),
+			                sysOrSame -> sysOrSame.booleanValue() ? Mono.just(sysOrSame)
+			                        : this.clientService.isBeingManagedBy(loggedInUserClientId, user.getClientId())
+			                                .flatMap(BooleanUtil::safeValueOfWithEmpty)
+			                                .flatMap(e -> SecurityContextUtil.hasAuthority("Authorities.User_UPDATE"))
+			                                .flatMap(BooleanUtil::safeValueOfWithEmpty))
+			                .switchIfEmpty(securityMessageResourceService.throwMessage(
+			                        msg -> new GenericException(HttpStatus.FORBIDDEN, msg),
+			                        SecurityMessageResourceService.HIERARCHY_ERROR));
 
-		        (ca, rclientId, isSameClient, isManaged) -> !isManaged.booleanValue() ? Mono.just(false)
-		                : SecurityContextUtil.hasAuthority("Authorities.User_UPDATE")
+		        }).contextWrite(Context.of(LogUtil.METHOD_NAME, "UserService.checkHierarchy"));
 
-		).contextWrite(Context.of(LogUtil.METHOD_NAME, "UserService.checkHierarchy"));
 	}
 
 	public Mono<Boolean> checkPasswordEquality(User u, RequestUpdatePassword reqPassword) {
 
 		if (u.isPasswordHashed())
 			return Mono.just(passwordEncoder.matches(u.getId() + reqPassword.getOldPassword(), u.getPassword())
-			        && passwordEncoder.matches(u.getId() + reqPassword.getNewPassword(), u.getPassword()));
+			        && !passwordEncoder.matches(u.getId() + reqPassword.getNewPassword(), u.getPassword()));
 
 		return Mono.just(StringUtil.safeEquals(reqPassword.getNewPassword(), u.getPassword())
```

Comment:
Need to add the ! here in this line above.

Reply:
updated the changes